gravitee-node-notifier/src/main/java/io/gravitee/node/notifier/trigger/NotificationTrigger.java

```diff
@@ -0,0 +1,215 @@
+  public void handle(Long event) {
+    if (condition.test(definition)) {
+      // keep EventLoop context in order to use it to execute the notifier's send method
+      final Context triggerContext = vertx.getOrCreateContext();
+      this.notificationAcknowledgeRepository.findByResourceIdAndAudienceId(
+          this.definition.getResourceId(),
+          definition.getType(),
+          definition.getAudienceId()
+        )
+        .map(Optional::ofNullable)
+        .switchIfEmpty(Maybe.just(Optional.empty()))
+        .subscribe(
+          notifAck -> {
+            if (notifAck.isEmpty()) {
+              // instantiate the plugin and send notification
+              final Optional<Notifier> optNotifier = notifierFactory.create(
+                definition
+              );
+              if (optNotifier.isPresent()) {
+                LOGGER.debug(
+                  "Send Notification with notifier type {} to audience {} about resource {}",
+                  definition.getType(),
+                  definition.getAudienceId(),
+                  definition.getResourceId()
+                );
+                final Notifier notifier = optNotifier.get();
+                // execute the send method into the triggerContext to avoid NullPointer as some notifiers rely on Vertx.currentContext
+                // that will be null as this block of code is executed in a RxJava ThreadPool and not into the Vertx EventLoop
+                triggerContext.runOnContext(
+                  aVoid ->
+                    notifier
+                      .send(buildNotification(definition), definition.getData())
+                      .whenComplete(
+                        (Void anotherVoid, Throwable throwable) -> {
+                          if (throwable != null) {
+                            LOGGER.error(
+                              "An error occurs while sending notification to {}",
+                              definition.getType(),
+                              throwable
+                            );
+                          } else {
+                            LOGGER.debug(
+                              "A notification has been sent to {}",
+                              definition.getType()
+                            );
+                            final NotificationAcknowledge notificationAcknowledge = new NotificationAcknowledge();
+                            notificationAcknowledge.setCreatedAt(new Date());
+                            notificationAcknowledge.setType(
+                              definition.getType()
+                            );
+                            notificationAcknowledge.setAudienceId(
+                              definition.getAudienceId()
+                            );
+                            notificationAcknowledge.setResourceId(
+                              definition.getResourceId()
+                            );
+                            notificationAcknowledgeRepository
```

Comment:
When multiple nodes will register the same notification trigger, there are good chances that the delay between the find ack and the create ack be big. The notification will be probably sent multiple times because of the cron (the process will be certainly executed at the exact same time) or because of the database replication delay.

Maybe something to discuss but a simple solution could be to apply a kind of random value when calculating the next tick ? (with something to explicitly activate when registering the trigger such as 'tryAvoidDuplicateNotification')

Reply:
why not simply rely on the database (mongo or RDBS) using a uniq index on the triplet (resource / type / aud )?

Reply:
Because if you have multiple instances of mgt all instances will register and run the same trigger on its side with the same identifier resource / type / aud right ? So each instance will try to find if an ack exists in the DB in the exact same time and, if not, will send the same notification twice (or N times if you have N instances). You know what I mean ?

Reply:
ok, I define a random value (between 0 and 10 seconds). let me know if ok for you